hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -570,21 +570,30 @@ HeartbeatTimeout detectTimeout(long heartbeatTimeoutMillis) {
             return NO_TIMEOUT__CALL_TIMEOUT_DISABLED;
         }
 
+        long invocationTime = op.getInvocationTime();
+
         // a call is always allowed to execute as long as its own call timeout
-        long deadlineMillis = op.getInvocationTime() + callTimeoutMillis;
-        if (deadlineMillis > context.clusterClock.getClusterTime()) {
+        if (invocationTime + callTimeoutMillis > context.clusterClock.getClusterTime()) {
             return NO_TIMEOUT__CALL_TIMEOUT_NOT_EXPIRED;
         }
 
-        // on top of its own call timeout, it is allowed to execute until there is a heartbeat timeout;
-        // so if the callTimeout is five minutes, and the heartbeatTimeout is one minute, then the operation is allowed
-        // to execute for at least six minutes before it is timing out
         long lastHeartbeatMillis = this.lastHeartbeatMillis;
-        long heartbeatExpirationTimeMillis = lastHeartbeatMillis == 0
-                ? op.getInvocationTime() + callTimeoutMillis + heartbeatTimeoutMillis
-                : lastHeartbeatMillis + heartbeatTimeoutMillis;
+        // It can happen that the lastHeartbeatMillis is from a previous retry. So instead of doing a 0 check,
+        // we check if the invocationTime is past the lastHeartbeatMillis
+        long heartbeatExpirationMillis;
+        if (invocationTime > lastHeartbeatMillis) {
+            // no sensible heartbeat was received.
+
+            // on top of its own call timeout, it is allowed to execute until there is a heartbeat timeout;
+            // so if the callTimeout is five minutes, and the heartbeatTimeout is one minute, then the operation is allowed
+            // to execute for at least six minutes before it is timing out
+            heartbeatExpirationMillis = invocationTime + callTimeoutMillis + heartbeatTimeoutMillis;
```

Comment:
I think there could be another problem here. InvocationTime is based on 'clusterClock', but we now use invocationTime in the heartbeatExpirationMillis; where it is compared with a regular Clock. This was already in the cde; not part of these changes.